File: 2463. minimum-total-distance-traveled.py
from typing import List
from collections import deque

class Solution:
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        # Sort positions to enable optimal matching
        robot.sort()
        factory.sort()
        
        m, n = len(robot), len(factory)
        # DP table initialization
        dp = [[0]*(n+1) for _ in range(m+1)]
        
        # Base case: if no factories left, distance is infinite
        for i in range(m): 
            dp[i][-1] = float('inf')
        
        # Process each factory from right to left
        for j in range(n-1, -1, -1):
            prefix = 0
            qq = deque([(m, 0)])
            
            # Process each robot from right to left
            for i in range(m-1, -1, -1):
                # Add distance to current factory
                prefix += abs(robot[i] - factory[j][0])
                
                # Remove elements outside factory limit window
                if qq[0][0] > i+factory[j][1]: 
                    qq.popleft()
                
                # Maintain monotonic queue property
                while qq and qq[-1][1] >= dp[i][j+1] - prefix:
                    qq.pop()
                    
                qq.append((i, dp[i][j+1] - prefix))
                dp[i][j] = qq[0][1] + prefix
                
        return dp[0][0]

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    robot1 = [0, 4, 6]
    factory1 = [[2, 2], [6, 2]]
    output1 = sol.minimumTotalDistance(robot1, factory1)
    print(f"Example 1: robot={robot1}, factory={factory1}")
    print(f"Output: {output1}")
    print(f"Expected: 4")
    print()
    
    # Example 2
    robot2 = [1, -1]
    factory2 = [[-2, 1], [2, 1]]
    output2 = sol.minimumTotalDistance(robot2, factory2)
    print(f"Example 2: robot={robot2}, factory={factory2}")
    print(f"Output: {output2}")
    print(f"Expected: 2")

File: 2463. minimum-total-distance-traveled.java
import java.util.*;

class Solution {
    long[][] dp;

    public long minimumTotalDistance(List<Integer> robot, int[][] factory) {
        Collections.sort(robot);
        Arrays.sort(factory, (a, b) -> a[0] - b[0]);

        int n = robot.size();
        int m = factory.length;

        dp = new long[n][m];
        for (long[] row : dp) Arrays.fill(row, -1);

        return solve(0, 0, robot, factory);
    }

    private long solve(int i, int j, List<Integer> robot, int[][] factory) {
        int n = robot.size();
        int m = factory.length;

        // all robots assigned
        if (i == n) return 0;

        // no factory left
        if (j == m) return (long)1e15;

        if (dp[i][j] != -1) return dp[i][j];

        long res = solve(i, j + 1, robot, factory); // skip factory

        long cost = 0;
        int pos = factory[j][0];
        int limit = factory[j][1];

        // assign k robots to this factory
        for (int k = 0; k < limit && i + k < n; k++) {
            cost += Math.abs(robot.get(i + k) - pos);
            res = Math.min(res, cost + solve(i + k + 1, j + 1, robot, factory));
        }

        return dp[i][j] = res;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        List<Integer> robot1 = Arrays.asList(0, 4, 6);
        int[][] factory1 = {{2, 2}, {6, 2}};
        long output1 = sol.minimumTotalDistance(robot1, factory1);
        System.out.println("Example 1: robot=[0,4,6], factory=[[2,2],[6,2]]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 4");
        System.out.println();
        
        // Example 2
        List<Integer> robot2 = Arrays.asList(1, -1);
        int[][] factory2 = {{-2, 1}, {2, 1}};
        long output2 = sol.minimumTotalDistance(robot2, factory2);
        System.out.println("Example 2: robot=[1,-1], factory=[[-2,1],[2,1]]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 2");
    }
}